### PR TITLE
Expose S3 and GCS `delimiter` parameter for listing objects

### PR DESCRIPTION
**Pull request by utkarsharma2**
# Description
## What is the current behavior?
Currently, we are not exposing the `delimiter` parameter for pattern resolution for S3 and GCS file locations. Which significantly limits the use of external files.

closes: #690 

## What is the new behavior?
Expose a `delimiter` which then is passed to the upstream API calls.

## Does this introduce a breaking change?
Nope

### Checklist
- [X] Created tests that fail without the change (if possible)

### CODE REVIEW

**Review by feluelle — COMMENTED** [review]
LGTM, but don't we need to mention it in the docs?

**Review by tatiana — COMMENTED** [review]
@utkarsharma2 please, could you add some tests to highlight the difference in the behaviour before and after this change?

**Review by sunank200 — COMMENTED** [review]
LGTM. Adding test for the test would be useful

File: src/astro/files/base.py [hunk 1]
```diff
@@ -24,22 +24,24 @@ def __init__(
         conn_id: str | None = None,
         filetype: constants.FileType | None = None,
         normalize_config: dict | None = None,
+        delimiter: str | None = None,
```
Comment (tatiana) [inline src/astro/files/base.py:27]: We'll probably need to align this with #772, depending on which PR is merged first.
Should we have a test to cover this new parameter support?

File: src/astro/files/base.py [hunk 2]
```diff
@@ -24,22 +24,24 @@ def __init__(
...
-        self.location: BaseFileLocation = create_file_location(path, conn_id)
+        self.location: BaseFileLocation = create_file_location(path, conn_id, delimiter)
         self.type = create_file_type(
             path=path, filetype=filetype, normalize_config=normalize_config
         )
 
     @property
     def path(self) -> str:
-        return self.location.path
+        return str(self.location.path)  # ToDo: look into typing issue
```
Comment (feluelle) [inline src/astro/files/base.py:44]: Could we try to fix that before merging? Let me know if you need help.

**Comment by utkarsharma2** [comment]
We don't need this now.

**Pull request closed without merging**